app/utils/marketplace_router.rb

```diff
@@ -160,37 +139,126 @@ def redirect_target(reason:, request:, community:, paths:, configs:, protocol:)
-      when :use_ident
+        { url: domain_redirect_url(domain: community[:domain], request: request),
+          status: :moved_permanently
+        }
+      when :use_ident, :www_ident
         # Community has a domain, but it's not in use.
         # -> Redirect to subdomain (ident)
-        {url: "#{protocol}#{community[:ident]}.#{configs[:app_domain]}#{request[:port_string]}#{request[:fullpath]}", status: :moved_permanently}
-      when :www_ident
+        # OR
         # Accessed community with ident, including www
         # -> Redirect to ident without www
-        {url: "#{protocol}#{community[:ident]}.#{configs[:app_domain]}#{request[:port_string]}#{request[:fullpath]}", status: :moved_permanently}
+        { url: ident_redirect_url(ident: community[:ident], app_domain: configs[:app_domain], request: request),
+          status: :moved_permanently
+        }
       else
         raise ArgumentError.new("Unknown redirect reason: '#{reason}'")
       end
 
     HashUtils.compact(DataTypes::Target.call(target.merge(reason: reason)))
   end
 
-  def redirect_reason(request:, community:, configs:, other:)
-    if other[:community_search_status] == :not_found && other[:no_communities]
+  def domain_redirect_url(domain:, request:)
+    host_redirect_url(host: domain, request: request)
+  end
+
+  def ident_redirect_url(ident:, app_domain:, request:)
+    host_redirect_url(host: "#{ident}.#{app_domain}", request: request)
+  end
+
+  def host_redirect_url(host:, request:)
+    "#{request[:protocol]}#{host}#{request[:port_string]}#{request[:fullpath]}"
+  end
+
+  # Returns a redirect reason or nil, if no redirect should be made
+  #
+  def redirect_reason(community:, host:, no_communities:, app_domain:)
+    community = Maybe(community).map { |c| DataTypes.create_community(c) }.or_else(nil)
+
+    if no_communities
       :no_marketplaces
-    elsif other[:community_search_status] == :not_found && !other[:no_communities]
+    elsif community.nil? && !no_communities
       :not_found
     elsif community && community[:deleted]
       :deleted
     elsif community && community[:closed]
       :closed
-    elsif community && community[:domain].present? && community[:use_domain] && request[:host] != community[:domain]
+    elsif community && community[:domain].present? && community[:use_domain] && host != community[:domain]
       :use_domain
-    elsif community && community[:domain].present? && !community[:use_domain] && request[:host] == community[:domain]
+    elsif community && community[:domain].present? && !community[:use_domain] && host == community[:domain]
       :use_ident
-    elsif community && request[:host] == "www.#{community[:ident]}.#{configs[:app_domain]}"
+    elsif community && host == "www.#{community[:ident]}.#{app_domain}"
       :www_ident
     end
   end
+
+  # This is a dirty method that is aware of the global application
```

Comment:
The word dirty in my mind refers to some data being modified but not flushed. I would call this side effecting in the comment text.